File: lib/utils/formatters.py
"""Pure formatting helpers — no side effects, no external calls."""
from datetime import date, datetime


def format_idr(amount: float) -> str:
    """Format number as Indonesian Rupiah string."""
    return f"Rp {amount:,.0f}".replace(",", ".")


def format_date(d: date | str, fmt: str = "%d %b %Y") -> str:
    """Format date to human-readable string."""
    if isinstance(d, str):
        d = datetime.fromisoformat(d).date()
    return d.strftime(fmt)


def format_pct(value: float, total: float) -> int:
    """Calculate percentage, safe division."""
    if total <= 0:
        return 0
    return int(value / total * 100)


def format_cycle_label(start: date, end: date) -> str:
    """Format cycle as '25 Mar - 24 Apr 2026'."""
    return f"{start.strftime('%d %b')} - {end.strftime('%d %b %Y')}"

File: webapp/assets/utils/formatters.js
/**
 * Format number as Indonesian Rupiah
 * @param {number} amount
 * @returns {string} e.g. "Rp 25.000"
 */
export function formatIDR(amount) {
  return 'Rp ' + Math.round(amount).toLocaleString('id-ID');
}

/**
 * Format date string to human-readable
 * @param {string} dateStr - YYYY-MM-DD
 * @returns {string} e.g. "7 Apr 2026"
 */
export function formatDate(dateStr) {
  if (!dateStr) return '';
  const d = new Date(dateStr + 'T00:00:00');
  return d.toLocaleDateString('id-ID', { day: 'numeric', month: 'short', year: 'numeric' });
}

/**
 * Format date relative to today
 * @param {string} dateStr
 * @returns {string} "Hari Ini" | "Kemarin" | "7 Apr 2026"
 */
export function formatDateRelative(dateStr) {
  const today = new Date().toISOString().slice(0, 10);
  const yesterday = new Date(Date.now() - 86400000).toISOString().slice(0, 10);
  if (dateStr === today) return 'Hari Ini';
  if (dateStr === yesterday) return 'Kemarin';
  return formatDate(dateStr);
}

/**
 * Format percentage safely
 * @param {number} value
 * @param {number} total
 * @returns {number}
 */
export function formatPct(value, total) {
  if (!total || total <= 0) return 0;
  return Math.round((value / total) * 100);
}
